455 favorite meal

priyaguntupalli5 opened the pull request:
![fav](https://github.com/CivicTechFredericton/mealplanner/assets/97687268/3cf0209a-d3b6-4f60-ab3a-8970f152e8f9)

![fav](https://github.com/CivicTechFredericton/mealplanner/assets/97687268/086b9284-5299-494b-90e4-c5f26c118823)

**Describe the technical changes contained in this PR**
Created a graphql fragment for favorite meals and added the selected favorite meal Id's to the gQLocalState. This local state is used throughout PersonFavoriteMeals, MealCard and Meals components. Every user has been assigned a slug and it is passed as an argument to the query, which results in favorite meals that are mapped to a particular user.

**Previous behaviour**
There is no option to query the favorites

**New behaviour**
User can now select and unselect their favorite meals and view them under favorites. 

**Related issues addressed by this PR**
Fixes #455 

**Have the following been addressed?**
- [ ] Have test cases been created for all of the changes?
- [ ] Do all of the test cases pass?
- [ ] Has the testing been done using the default docker-compose environment?
- [ ] Are documentation changes required?
- [ ] Does this change break or alter existing behaviour?

priyaguntupalli5 committed:
favorite meal plan feature

shanthisa commented on backend/015-favorite-meal-plan.sql:
```diff
@@ -0,0 +1,30 @@
+-- Create the favorite_meal_plan table if it doesn't exist
+CREATE TABLE IF NOT EXISTS app.favorite_meal_plan (
```
There seems to be a confusion of favorite meals and favorite meal plans. We are only addressing favorite meals. So this should be called favorite_meals. 

To create a favorite meal, you don't need to pass the user_id. Just the meal_id is sufficient. The current user logged in should be retrieved from the `jwt.claims.person_id` which can be gotten from `current_setting`.

Basically the mutations can be just with the argument mealid like below:

```
mutation {
  createFavoriteMeal(mealId: "id") {
    // this returns meal
  }
}
```
Also we need to delete the favorite meal from the table, when it is favorite icon is deselected.
```
mutation {
  deleteFavoriteMeal(mealId: "1") {
    // attributes of the meal.
  }
}
```

For the query, you must get favorite meals for a logged in user.

```
query {
  favoriteMeals {
    // attributes for the meal.
  }
}
```

We need to create two functions for creating and removing the favorite meal. We also need another function to return the favoriteMeals of the current user. 

Potentially, in future, we can have a feature where a meal designer can know how favorite a meal is by retrieving the users who have favoritied a meal.

```
query {
  meal(id: "id") {
    // other attributes
    favoritedBy {
      nodes {
          fullName
          email
           id
       }
    }
  }
}
```

priyaguntupalli5 committed:
updated sql file

priyaguntupalli5 committed:
favorite meals

priyaguntupalli5 committed:
rebase with v2

priyaguntupalli5 committed:
granted execute access

shanthisa commented on mealplanner-ui/src/App.tsx:
```diff
@@ -108,16 +108,15 @@ function App() {
             />
-            <Route 
-              path={`/meals/:slug/favorites`}
+            {/* <Route path={`/meals/:slug/favorites`}
```
Why is the route path for favorites commented?

shanthisa commented on mealplanner-ui/src/pages/Meals/MealCard.tsx:
```diff
@@ -20,16 +26,46 @@ interface ExpandMoreProps extends IconButtonProps {
       setExpanded(!expanded);
     };
+
+    const handleToogleFavorite = (id: string) => {
```
Spelling mistake Toggle

shanthisa commented on mealplanner-ui/src/pages/Meals/Meals.tsx:
```diff
@@ -34,20 +36,29 @@ const mealsQuery = graphql`
     { fetchPolicy: "store-or-network" }
   );
 
+  const PFMeals: any = useRefetchableFragment(FavoriteMealsFragment, data)[0];
```
Should not give type any. Please give valid type

shanthisa commented on mealplanner-ui/src/pages/Meals/Meals.tsx:
```diff
@@ -34,20 +36,29 @@ const mealsQuery = graphql`
+  const PFMeals: any = useRefetchableFragment(FavoriteMealsFragment, data)[0];
+  const [_, refetch] = useRefetchableFragment(FavoriteMealsFragment, data);
   const selectedTags = data.gqLocalState.selectedMealTags || [];
+  const selectedFavs = PFMeals.people?.nodes[0].favoriteMeals.nodes.map((favMeal:any) => favMeal.meal?.rowId) || [];
```
favMeal type should be defined properly. Not as any

shanthisa commented on mealplanner-ui/src/pages/Meals/Meals.tsx:
```diff
@@ -34,20 +36,29 @@ const mealsQuery = graphql`
 export const Meals = () => {
   const [searchMeal, setSearchMeal] = useState<string>("");
   const [searchType, setSearchType] = useState('name');
+  const slug = getCurrentPerson().personSlug;
```
For admin and meal designer, they should be able to select the user and view their favorite meals. Only for normal user it is current person.

shanthisa reviewed (changes requested):
As per the changes in #516, when a route is presented with a slug, depending on the user slug, the favorites is displayed.
This permission is given for admin and meal designer. 
This has been clearly described in the PR
https://github.com/CivicTechFredericton/mealplanner/pull/524
Your UI is not satisfying the backend of this. You are only doing the current user.  Please read the scope of 524 PR description and accordingly modify the frontend to support it.

As an admin or meal designer I should be able to view the favorite meals of any user. 
As a app_user I will be able to view only my current favorite meal. 

Your PR is not satisfying the requirements.

priyaguntupalli5 committed:
fixed types and spelling

priyaguntupalli5 committed:
added users to admin and mealdesigner

priyaguntupalli5 committed:
routing favorite meals

priyaguntupalli5 committed:
backend route fix

priyaguntupalli5 committed:
backend fix

shanthisa commented on backend/db_migrations/000006_favorite_meals.up.sql:
```diff
@@ -55,5 +55,7 @@ $$ language plpgsql;
 -- Grant execute permission to appropriate roles
 grant select, insert, delete, update on app.favorite_meals to app_user, app_meal_designer, app_admin;
 grant usage on sequence app.favorite_meals_id_seq to app_user, app_meal_designer, app_admin;
+grant execute on function app.add_favorite_meal(bigint) to app_user,app_admin,app_meal_designer;
```
You cannot edit existing migrations as it will not run on a database where migration already happened like our test server. You need to create a new migration.

shanthisa commented on backend/db-reset.psql:
```diff
@@ -13,4 +13,4 @@ drop schema if exists app_private, app cascade;
 \i 012-duplicate-meal-plan.sql
 \i 013-add-email-and-role-to-person.sql
-\i 014-reset-password.sql
+\i 014-reset-password.sql
```
This file should not have been touched.

shanthisa replied:
This is still not resolved

shanthisa commented on backend/migrations/006-516-favorite-meal.sql:
```diff
@@ -54,5 +54,7 @@ $$ LANGUAGE plpgsql;
 -- Grant execute permission to appropriate roles
 GRANT select, insert, delete, update on app.favorite_meals to app_user, app_meal_designer, app_admin;
 GRANT usage on SEQUENCE app.favorite_meals_id_seq to app_user, app_meal_designer, app_admin;
+GRANT execute on function app.add_favorite_meal(bigint) to app_user,app_admin,app_meal_designer;
```
Should not touch this file as this will be deleted in future as we moved migrations to db_migrations

shanthisa commented on mealplanner-ui/src/pages/Meals/AddFavoriteMeal.tsx:
```diff
@@ -0,0 +1,31 @@
+import { graphql } from "babel-plugin-relay/macro";
```
This should be a .ts file and not .tsx as there are no React components in this file.

shanthisa commented:
Also the UI should not have a line running through 'Select any User'. You need to use sx settings to modify the css and show the white space behind 'Select any User' instead of line striking through.

priyaguntupalli5 committed:
fixed label styling

shanthisa reviewed (changes requested):
Need the following changes

shanthisa commented:
Also there is a conflict in your branch for the MealsQuery.graphql.ts. You may want first to accept the v2 version. Then regenerate and see whether there are any changes.

priyaguntupalli5 committed:
renamed file extension

priyaguntupalli5 committed:
added migration file

shanthisa commented on mealplanner-ui/src/pages/Meals/Meals.tsx:
```diff
@@ -3,64 +3,69 @@ import {
-        code
-        photoUrl
-        videoUrl
+type FavoriteMeals = {
```
This type should not be explicitly written this way because when the type changes in the backend, you need to maintain this type. The better way to do is to extract it from relay types. Please refer types.ts under state folder. Please update the type similar to what is written with Exclude.

shanthisa commented on mealplanner-ui/src/pages/Meals/PersonFavoriteMeals.tsx:
```diff
@@ -36,20 +43,19 @@ const favoriteMealsQuery = graphql`
 `;
 
 export const FavoriteMeals = () => {
+  const favs = MealsData() as PersonFavoriteMeals_favorites$key;
```
This database query should not be re-run. Instead this should be passed as props from the parent component where this hook is already run.

shanthisa commented on mealplanner-ui/src/pages/Meals/MealsData.tsx:
```diff
@@ -0,0 +1,43 @@
+  }
+`;
+
+export const MealsData = () => {
```
This is a hook. So it should be named as useMealsData

shanthisa committed:
Added wrapper query to wrap the fragment
Added a new page component FavoritePage
Segregated the favorites of other users and My favorites
Created a navigation back to meals as the endpoint URL is different with slug
Removed unnecessary code such as useEffect and state variables and double data query
Added heading
Included person fullName to both heading
Modified no meals message

shanthisa commented:
As there was a rebase required and there were significant changes to be done, I implemented those and pushed it in PR #630
So closing this PR.
